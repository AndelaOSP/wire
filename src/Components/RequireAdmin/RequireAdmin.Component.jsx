import React from 'react';
import { withRouter } from 'react-router';
import { Redirect } from 'react-router-dom';
import PropTypes from 'prop-types';
import authenticateUser from '../../helpers/auth';

const RequireAdmin = Component => {
  class AuthenticatedComponent extends React.Component {
    render() {
      return authenticateUser.isAdmin() ? (
        <Component {...this.props} />
      ) : (
        <Redirect to={{ pathname: '/dashboard', state: { from: this.props.location } }} />
      );
    }
  }

  AuthenticatedComponent.propTypes = {
    location: PropTypes.object
  };

  return withRouter(AuthenticatedComponent);
};

/**
 * RequireAdmin PropTypes declaration
 */
RequireAdmin.propTypes = {
  component: PropTypes.oneOfType([PropTypes.node, PropTypes.func]) // Container || functional react component
};

export default RequireAdmin;
