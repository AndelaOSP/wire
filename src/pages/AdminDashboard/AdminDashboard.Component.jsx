import React, { Component } from 'react';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

// actions
import { fetchStaff, inviteUser, searchUsers, updateUser, removeUser } from '../../actions/staffAction';
import { fetchRoles } from '../../actions/rolesAction';
import { fetchLocations } from '../../actions/locationsAction';

// helpers
import { matchPositionToRoleId, matchLocationToLocationId } from '../../helpers/adminDashboard';

// styling
import './AdminDashboard.scss';

// Components
import NavBar from '../../Common/NavBar/NavBar.Component';
import AvailableUser from '../../Components/AvailableUser/AvailableUser.Component';
import UserFilter from '../../Components/UserFilter/UserFilter.Component';
import CustomNotification from '../../Components/CustomNotification/CustomNotification.Component';
import CircularProgressIndicator from '../../Components/Progress/Progress.Component';

/**
 * @class Dashboard
 */
export class AdminDashboard extends Component {
  constructor(props) {
    super(props);
    this.state = {
      countryFilter: 'All Countries',
      selectedUser: null,
      loggedInUser: localStorage.getItem('email')
    };
  }

  componentDidMount() {
    this.props.fetchStaff();
    this.props.fetchRoles();
    this.props.fetchLocations();
  }

  changeCountryFilter() {
    return key => {
      this.setState({ countryFilter: key });
    };
  }

  handleSearch = query => {
    if (query) {
      this.props.searchUsers(query);
    }
  };

  /**
   * Method to handle the invite user api call
   */
  handleInvite = (email, position, location) => {
    let { roles, locations } = this.props;
    let roleId = matchPositionToRoleId(roles, position);
    let locationId = matchLocationToLocationId(locations, location);
    this.props.inviteUser(email, roleId, locationId);
  };

  /**
   * Method to handle position change
   */
  handlePositionChange = (position, userId, index) => {
    let { roles } = this.props;
    let roleId = matchPositionToRoleId(roles, position);
    this.props.updateUser(userId, roleId, index);
  };

  /**
   * Method to handle user removal
   */
  handleRemove = (userId, index) => {
    this.props.removeUser(userId, index);
  };

  filterStaff = () => {
    let staff = this.props.staff;
    let {loggedInUser} = this.state;
    // filter by countries
    if (this.state.countryFilter !== 'All Countries') {
      staff = staff.filter(user => {
        return this.state.countryFilter.toLocaleLowerCase() === user.Location.country.toLowerCase();
      });
    }
    const newStaff = staff.filter(user => {
      return user.email !== loggedInUser;
    });
    return newStaff;
  };

  render() {
    const { isLoading, isError, errorMessage } = this.props;
    const staff = this.filterStaff();
    return (
      <div>
        <NavBar {...this.props} showSearch={false} />
        {isLoading ? (
          <CircularProgressIndicator />
        ) : (
          <div className="admin-dashboard">
            <UserFilter
              staff={staff}
              handleSearch={this.handleSearch}
              handleInvite={this.handleInvite}
              changeCountryFilter={this.changeCountryFilter()}
            />
            <div className="available-users">
              {staff.length ? (
                staff.map((staffMember, i) => (
                  <AvailableUser
                    key={i}
                    id={staffMember.id}
                    index={this.props.staff.indexOf(staffMember)}
                    imageUrl={staffMember.imageUrl}
                    username={staffMember.username}
                    role={staffMember.Role.name.toUpperCase()}
                    country={staffMember.Location.country}
                    handlePositionChange={this.handlePositionChange}
                    handleRemove={this.handleRemove}
                  />
                ))
              ) : (
                <div className="no-users">
                  <p>Sorry, no users. Invite some?</p>
                </div>
              )}
            </div>
          </div>
        )}
        {isError ? (
          <CustomNotification type={'error'} message={errorMessage} autoHideDuration={15000} open />
        ) : (
          <CustomNotification type={'error'} message={errorMessage} open={false} />
        )}
      </div>
    );
  }
}

/**
 * Dashboard Component Props validation
 */
AdminDashboard.propTypes = {
  isLoading: PropTypes.bool.isRequired,
  isError: PropTypes.bool.isRequired,
  errorMessage: PropTypes.string.isRequired,
  fetchStaff: PropTypes.func.isRequired,
  fetchRoles: PropTypes.func.isRequired,
  fetchLocations: PropTypes.func.isRequired,
  inviteUser: PropTypes.func.isRequired,
  searchUsers: PropTypes.func.isRequired,
  updateUser: PropTypes.func.isRequired,
  removeUser: PropTypes.func.isRequired,
  staff: PropTypes.array,
  roles: PropTypes.array,
  locations: PropTypes.array
};

/**
 * map state from the store to props
 * @param {*} state
 * @returns {*} partial state
 */
const mapStateToProps = state => {
  return {
    isLoading: state.isLoading,
    isError: state.error.status,
    errorMessage: state.error.message,
    staff: state.staff,
    roles: state.roles,
    locations: state.locations
  };
};

/**
 * map dispatch to props
 * @param {*} dispatch
 */
const mapDispatchToProps = dispatch =>
  bindActionCreators(
    {
      fetchStaff,
      fetchRoles,
      fetchLocations,
      inviteUser,
      searchUsers,
      updateUser,
      removeUser
    },
    dispatch
  );

export default connect(mapStateToProps, mapDispatchToProps)(AdminDashboard);
