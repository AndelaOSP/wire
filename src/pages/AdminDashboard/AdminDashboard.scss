@import '../../theme.scss';

.admin-dashboard {
  background-color: white;
  height: 91.5vh;
  width: 100vw;
}

.available-users {
  flex-direction: column;
  align-items: flex-start;
  height: 47.5vh;
  position: relative;
  top: 6vh;
  left: 7vw;
  width: 75vw;
  overflow: scroll;
}

.no-users {
  font-family: 'DIN Pro Extra Light';
  text-align: center;
}
