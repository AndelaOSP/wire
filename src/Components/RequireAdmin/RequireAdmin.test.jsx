import React from 'react';
import { shallow } from 'enzyme';
import shallowToJSON from 'enzyme-to-json';
import RequireAdmin from './RequireAdmin.Component';
import authenticateUser from '../../helpers/auth';
import { AdminDashboard } from '../../pages/AdminDashboard/AdminDashboard.Component';

jest.mock('../../helpers/auth');

authenticateUser.isAdmin = () => true;
authenticateUser.isAuthenticated = () => true;

describe('RequireAdmin Component', () => {
  it('should render without crashing', () => {
    const requireAdmin = shallow(React.createElement(RequireAdmin(AdminDashboard)));
    const tree = shallowToJSON(requireAdmin);
    expect(tree).toMatchSnapshot();
  });
});
