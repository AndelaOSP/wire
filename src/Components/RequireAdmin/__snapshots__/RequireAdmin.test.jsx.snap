// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RequireAdmin Component should render without crashing 1`] = `
<Route>
  <Component />
</Route>
`;
